src/lib/OccAnalysis.hs

```diff
@@ -462,7 +471,14 @@ instance HasOCC (ComposeE PrimOp Atom) where
     valIx <- summary val
     -- Treat the combining function as inlined here and called once
     void $ oneShot2 accessOnce (Deterministic []) valIx combine
-    -- Do not change it
+    -- Do not change the expression
+    -- TODO(correctness) There is a potential bug here.  If the combining
+    -- function contains some internally dead code, it both won't be removed by
+    -- this traversal and won't be counted as referencing its free variables.
+    -- This could then lead us to make a bad inlining decision about said free
+    -- variables.  Unless I am mistaken, that requires the monoid instance to
+    -- have been defined nested, which I think is rare enough that I'm inclined
+    -- let the possible bug slide until it occurs.
```

Comment:
Letting the bug slide until it occurs seems like a bad idea to me. Is there a check we can do that triggers a "not implemented" assertion? That would be an acceptable workaround for me.

Reply:
Actually, I think I found a reasonable way to fix the bug, and did so.